### ISSUE TITLE
Ability to Provide Title to Legend/ Legend Groups

### ISSUE THREAD

**Issue by alexander-alvarez** [plotly/plotly.js#689]
I'm :cry: that I can add a heading to the different groups generated by `legendgroup`.

I'd like to group my traces by a particular property on the input data, and also display a label.

Imagine data:

``` js
[{ 'group': 'A', name: 'thing 1'}, 
{ 'group': 'A', name: 'thing 2'}, 
{ 'group': 'B', name: 'thing' }]
```

If I want the legend to include context about the group, I'd have to prefix the name with the group:

```
- A: thing 1
- A: thing 2

- B: thing
```

Whereas if we had the ability to add an overarching legend headings (and for each group) we could do this:

```
A
 - thing 1
 - thing 2
B
 - thing

```

you can see where this is an issue with longer group names

**Comment by etpinard** [comment]
That would be nice! 

It won't be a priority for us though as legend headings can easily be mocked using with properly places annotations.

That said, it would a great opportunity for a community contributor.

The patch should be in `src/components/legend/draw.js`. Currently the has-legend-group logic is in this [block](https://github.com/plotly/plotly.js/blob/b6fc59d33b95ad52b6a83957807120f485c7be8a/src/components/legend/draw.js#L505-L507).

**Comment by bluprince13** [comment]
Wondered if there's been any progress on this. I posted a [question ](https://stackoverflow.com/questions/45555266/plotly-legend-title)on stackoverflow, and got a couple of good short term solutions, but it'd be good to have it native in the future.

**Comment by daattali** [comment]
Legend headings **can** be hacked using annotations, but it's far from an ideal solution. Getting the perfectly right is tough, and even worse is the fact that if the plotly doesn't account for space taken by the heading, so if the title is longer than the legend labels, it will run off the screen.

Would be awesome to at least support a basic title without groupings as a first step

**Comment by seansfkelley** [comment]
AFAICT, using absolute positioning annotations doesn't work if you have legend groups with varying lengths. I have two legend groups of unknown size, so I can't safely position a label at all times so as to be readable and not misleading. Unfortunately, that makes the legend significantly less useful, since many of the traces are named similarly. Unless there's a different workaround for this case?

**Comment by Braintelligence** [comment]
Also looking for a way to get this working if your plot varies in height/width in some situations (different plot layout for downloads or varying screen size).
I already have to account for the legend not clipping on 360px wide screens (still the mostly used width for smartphones globally), but hand-placing legend group titles for every occasion is just too much...

If we could somehow anchor annotations to legend groups this specific problem would be instantly solved.

**Comment by strass** [comment]
That plotly.py issue above says it was added in v4.5. Is there a similar feature for JS? I have the same issue as sean where my legend groups have variable sizes

**Comment by nicolaskruchten** [comment]
This issue has been tagged with `NEEDS SPON$OR`

A [community PR](https://github.com/plotly/plotly.js/pulls?q=is%3Apr+is%3Aopen+label%3A%22type%3A+community%22) for this feature would certainly be welcome, but our experience is deeper features like this are difficult to complete without the Plotly maintainers leading the effort.

**Sponsorship range: $10k-$15k**

What Sponsorship includes:
- Completion of this feature to the Sponsor's satisfaction, in a manner coherent with the rest of the Plotly.js library and API
- Tests for this feature
- Long-term support (continued support of this feature in the latest version of Plotly.js)
- Documentation at [plotly.com/javascript](https://plotly.com/javascript/)
- Possibility of integrating this feature with Plotly Graphing Libraries (Python, R, F#, Julia, MATLAB, etc)
- Possibility of integrating this feature with Dash
- Feature announcement on [community.plotly.com](https://community.plotly.com/tag/announcements) with shout out to Sponsor (or can remain anonymous)
- Gratification of advancing the world's most downloaded, interactive scientific graphing libraries (>50M downloads across supported languages)

Please include the link to this issue when [contacting us](https://plotly.com/get-pricing/) to discuss.

**Comment by rithwikjc** [comment]
I was looking for exactly the same thing @alexander-alvarez  was looking for. That is,
> ```
> A
>  - thing 1
>  - thing 2
> B
>  - thing
> ```
I used some hack to achieve a semi-desirable result. For me each group shares the same color, and the traces within the group have different `marker` or `dash` property. 
![Result](https://user-images.githubusercontent.com/25515417/121154992-0a00aa80-c865-11eb-9d0c-33079b493387.PNG)
I did this by adding a dummy trace, which essentially contains no data (and hence no effect on the plot).
```python
 fig.add_trace(go.Scatter(x=[0], y=[0], name='GroupTitle', legendgroup='GroupTitle',
                                     line_color=color_of_group , mode='lines',  line_width=10))
```
This essentially creates a new item in legend, but has it's name as the group name. The big block (line of width 10, represents the color of the group for me. 

You can also remove the block next to the group title by changing `line_width=0` or `opacity=0`, creating the legend as shown below. However I prefer a solid color block instead of a misaligned group title.
![Capture](https://user-images.githubusercontent.com/25515417/121155997-f144c480-c865-11eb-9de9-8f6bb7dc5784.PNG)

**Comment by nicolaskruchten** [comment]
A short update on this issue: it covers legend titles (which was implemented last year) and legend *group* titles which we are working on now, tracked in https://github.com/plotly/plotly.js/issues/5260

**Comment by nicolaskruchten** [comment]
We received anonymous sponsorship for this item and implemented legend group titles in  https://github.com/plotly/plotly.js/pull/5752 (to be released in v2.2). Clickable legend group titles are now tracked separately here: https://github.com/plotly/plotly.js/issues/5765 and separately toggle-able grouped items are tracked here https://github.com/plotly/plotly.js/issues/3135

Finally, it seems is if part of the desire to group legend items with titles stems from actually wanting multiple separate, titled legends. We added legend titles last year, and the idea of having multiple independently-positionable legends is tracked here https://github.com/plotly/plotly.js/issues/5098.

I'm closing this issue as 'resolved' and locking it for now and encouraging folks with follow-up or related comments to use one of the issues above please :)
